### PR TITLE
Lock directories to synchronize wheel-install copies

### PR DESCRIPTION
**Pull request by charliermarsh**
## Summary

Closes https://github.com/astral-sh/uv/issues/4831.

### CODE REVIEW

File: crates/install-wheel-rs/src/linker.rs [hunk 1]
```diff
@@ -536,3 +547,23 @@ fn hardlink_wheel_files(
...
+/// Copy from `from` to `to`, ensuring that the parent directory is locked.
+fn synchronized_copy(from: &Path, to: &Path, locks: &Locks) -> std::io::Result<()> {
+    // Ensure we have a lock for the directory.
+    let dir_lock = {
+        let mut locks_guard = locks.0.lock().unwrap();
+        locks_guard
+            .entry(to.parent().unwrap().to_path_buf())
+            .or_insert_with(|| Arc::new(Mutex::new(())))
+            .clone()
+    };
```
Comment (charliermarsh) [inline crates/install-wheel-rs/src/linker.rs:560]: @konstin - did you have some idea of only using a lock per _top-level_ directory here? Why do that?

File: crates/install-wheel-rs/src/linker.rs [hunk 2]
```diff
@@ -25,6 +27,9 @@ use crate::wheel::{
 };
 use crate::{Error, Layout};
 
+#[derive(Debug, Default)]
+pub struct Locks(Arc<Mutex<FxHashMap<PathBuf, Arc<Mutex<()>>>>>);
```
Comment (ibraheemdev) [inline crates/install-wheel-rs/src/linker.rs:31]: Looks like the outer `Arc` isn't necessary here.

This also looks very similar to the structure in `uv-distribution`, wonder if we could reuse it somehow.

File: crates/install-wheel-rs/src/linker.rs (hunk 1 above)
Comment (zanieb) [reply crates/install-wheel-rs/src/linker.rs:563]: I think the idea is to hold the lock longer?

> We keep that lock until we have a new directory, so when the next files are `foo/bar/b.py` and `foo/baz/c.py`, we already have the lock and it's just a prefix check.

Comment (konstin) [reply crates/install-wheel-rs/src/linker.rs:563]: Yeah, it was a micro-optimization

**Review by konstin — APPROVED** [review]

**Pull request merged**